Algorithm/QCAlgorithm.History.cs

```diff
@@ -624,34 +626,91 @@ public BaseData GetLastKnownPrice(Security security)
             //       The actual issue is Slice.GetImpl, so patch this can be removed right after it is properly addressed.
             //       A proposed solution making the Tick class a BaseDataCollection and make the Ticks class a dictionary Symbol->Tick instead of
             //       Symbol->List<Tick> so we can use the Slice.Get methods to collect all ticks in every slice instead of only the last one.
-            if (typeof(T) == typeof(Tick))
+            if (type == typeof(Tick))
             {
                 return (IEnumerable<T>)slices.Select(x => x.Ticks).Where(x => x.ContainsKey(symbol)).SelectMany(x => x[symbol]);
             }
 
-            return slices.Get<T>(symbol).Memoize();
+            return slices.Get<T>(symbol);
+        }
+
+        /// <summary>
+        /// Centralized logic to get data typed history for a given list of requests.
+        /// </summary>
+        /// <remarks>
+        /// This method will check for Python custom data types in order to call the right Slice.Get dynamic method
+        /// </remarks>
+        private IEnumerable<DataDictionary<T>> GetDataTypedHistory<T>(IEnumerable<HistoryRequest> requests)
+            where T : IBaseData
+        {
+            var slices = History(requests.Where(x => x != null));
+
+            if (typeof(T) == typeof(PythonData))
+            {
+                return GetPythonCustomDataTypeHistory(slices, requests.First().DataType).OfType<DataDictionary<T>>().Memoize();
+            }
+
+            return slices.Get<T>();
+        }
+
+        /// <summary>
+        /// Centralized logic to get data typed history given a list of requests for the specified symbol.
+        /// This method is used to keep backwards compatibility for those History methods that expect an ArgumentException to be thrown
+        /// when the security and the requested data type do not match
+        /// </summary>
+        /// <remarks>
+        /// This method is only used for Python algorithms, specially for those requesting custom data type history.
+        /// The reason for using this method is that custom data type Python history calls to
+        /// <see cref="History{T}(QuantConnect.Symbol, int, Resolution?)"/> will always use <see cref="PythonData"/> (the custom data base class)
+        /// as the T argument, because the custom data class is a Python type, which will cause the history data in the slices to not be matched
+        /// to the actual requested type, resulting in an empty list of slices.
+        /// </remarks>
+        private IEnumerable<dynamic> GetPythonCustomDataTypeHistory(IEnumerable<Slice> slices, Type pythonType, Symbol symbol = null)
+        {
+            var history = slices.Select(x => x.Get(pythonType));
+
+            if (symbol == null)
+            {
+                return history;
+            }
+
+            return history.Where(x => x.ContainsKey(symbol)).Select(x => x[symbol]);
```

Comment:
Maybe some of these could go into the `SliceExtensions` along side the other Get implementations? Potentially there could be a single method which covers both use cases and we directly call that one?